src/pss.rs

```diff
@@ -0,0 +1,323 @@
+    salt_len: Option<usize>,
+    digest: &mut dyn DynDigest,
+) -> Result<Vec<u8>> {
+    let salt_len = salt_len.unwrap_or_else(|| priv_key.size() - 2 - digest.output_size());
+
+    let mut salt = vec![0; salt_len];
+    rng.fill(&mut salt[..]);
+
+    sign_pss_with_salt(blind_rng, priv_key, hashed, &salt, digest)
+}
+
+/// signPSSWithSalt calculates the signature of hashed using PSS [1] with specified salt.
+/// Note that hashed must be the result of hashing the input message using the
+/// given hash function. salt is a random sequence of bytes whose length will be
+/// later used to verify the signature.
+fn sign_pss_with_salt<T: Rng, SK: PrivateKey>(
+    blind_rng: Option<&mut T>,
+    priv_key: &SK,
+    hashed: &[u8],
+    salt: &[u8],
+    digest: &mut dyn DynDigest,
+) -> Result<Vec<u8>> {
+    let em_bits = priv_key.n().bits() - 1;
+    let mut em = vec![0; (em_bits + 7) / 8];
+    emsa_pss_encode(&mut em, hashed, em_bits, salt, digest)?;
+
+    priv_key.raw_decryption_primitive(blind_rng, &em, priv_key.size())
+}
+
+fn emsa_pss_encode(
+    em: &mut [u8],
+    m_hash: &[u8],
+    em_bits: usize,
+    salt: &[u8],
+    hash: &mut dyn DynDigest,
+) -> Result<()> {
+    // See [1], section 9.1.1
+    let h_len = hash.output_size();
+    let s_len = salt.len();
+    let em_len = (em_bits + 7) / 8;
+
+    // 1. If the length of M is greater than the input limitation for the
+    //     hash function (2^61 - 1 octets for SHA-1), output "message too
+    //     long" and stop.
+    //
+    // 2.  Let mHash = Hash(M), an octet string of length hLen.
+    if m_hash.len() != h_len {
+        return Err(Error::InputNotHashed);
+    }
+
+    // 3. If em_len < h_len + s_len + 2, output "encoding error" and stop.
+    if em_len < h_len + s_len + 2 {
+        // TODO: Key size too small
+        return Err(Error::Internal);
+    }
+
+    if em.len() != em_len {
+        return Err(Error::Internal);
+    }
+
+    let (db, h) = em.split_at_mut(em_len - s_len - h_len - 2 + 1 + s_len);
+    let h = &mut h[..(em_len - 1) - db.len()];
+
+    // 4. Generate a random octet string salt of length s_len; if s_len = 0,
+    //     then salt is the empty string.
+    //
+    // 5.  Let
+    //       M' = (0x)00 00 00 00 00 00 00 00 || m_hash || salt;
+    //
+    //     M' is an octet string of length 8 + h_len + s_len with eight
+    //     initial zero octets.
+    //
+    // 6.  Let H = Hash(M'), an octet string of length h_len.
+    let prefix = [0u8; 8];
+
+    hash.input(&prefix);
+    hash.input(m_hash);
+    hash.input(salt);
+
+    let hashed = hash.result_reset();
+    h.copy_from_slice(&hashed);
+
+    // 7.  Generate an octet string PS consisting of em_len - s_len - h_len - 2
+    //     zero octets. The length of PS may be 0.
+    //
+    // 8.  Let DB = PS || 0x01 || salt; DB is an octet string of length
+    //     emLen - hLen - 1.
+    db[em_len - s_len - h_len - 2] = 0x01;
+    db[em_len - s_len - h_len - 1..].copy_from_slice(salt);
```

Comment:
We're assuming here that the `em` buffer passed into `emsa_pss_encode` is already zeroed. This happens to be the case, but given that `emsa_pss_encode` is only called from one place, and this PR isn't attempting to be no-std compatible, it would be clearer if `em` was constructed and returned from this function. Non-blocking.

Reply:
fixed